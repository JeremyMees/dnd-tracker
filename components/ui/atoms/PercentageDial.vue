<script setup lang="ts">
const props = defineProps<{
  limit: number
  value: number
}>()

const reversedPercentage = computed<number>(() => {
  const value = Math.round((100 / props.limit) * props.value)
  return value - 100
})
</script>

<template>
  <svg
    class="size-7 -rotate-90"
    viewBox="0 0 36 36"
  >
    <circle
      cx="18"
      cy="18"
      r="16"
      fill="none"
      stroke-width="4"
      class="stroke-current text-slate-700"
    />
    <circle
      cx="18"
      cy="18"
      r="16"
      fill="none"
      stroke-width="4"
      stroke-dasharray="100"
      :stroke-dashoffset="reversedPercentage"
      stroke-linecap="round"
      :class="value === limit ? 'text-danger' : 'text-primary'"
      class="stroke-current duration-300 transition-all"
    />
  </svg>
</template>
