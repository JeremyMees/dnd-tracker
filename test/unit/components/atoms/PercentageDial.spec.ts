import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'
import PercentageDial from '~/components/atoms/PercentageDial.vue'

interface Props {
  limit: number
  value: number
}

const props: Props = {
  limit: 100,
  value: 50,
}

describe('PercentageDial', async () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(PercentageDial, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should set stroke-dashoffset to the correct value', async () => {
    const component = await mountSuspended(PercentageDial, { props })
    const dial = component.find('[data-test-dial]')

    expect(dial.attributes('stroke-dashoffset')).toBe('-50')

    component.setProps({ value: 30 })
    await nextTick()

    expect(dial.attributes('stroke-dashoffset')).toBe('-70')
  })
})
