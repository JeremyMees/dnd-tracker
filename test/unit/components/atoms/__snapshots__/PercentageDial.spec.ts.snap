// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PercentageDial > Should match snapshot 1`] = `
"<svg class="size-7 -rotate-90" viewBox="0 0 36 36">
  <circle cx="18" cy="18" r="16" fill="none" stroke-width="4" class="stroke-current text-secondary"></circle>
  <circle data-test-dial="" cx="18" cy="18" r="16" fill="none" stroke-width="4" stroke-dasharray="100" stroke-dashoffset="-50" stroke-linecap="round" class="text-primary stroke-current duration-300 transition-all"></circle>
</svg>"
`;
